pkg/domain/infra/abi/system.go

```diff
@@ -390,13 +390,25 @@ func unshareEnv(graphroot, runroot string) []string {
 		fmt.Sprintf("CONTAINERS_RUNROOT=%s", runroot))
 }
 
-func (ic *ContainerEngine) Unshare(ctx context.Context, args []string) error {
-	cmd := exec.Command(args[0], args[1:]...)
-	cmd.Env = unshareEnv(ic.Libpod.StorageConfig().GraphRoot, ic.Libpod.StorageConfig().RunRoot)
-	cmd.Stdin = os.Stdin
-	cmd.Stdout = os.Stdout
-	cmd.Stderr = os.Stderr
-	return cmd.Run()
+func (ic *ContainerEngine) Unshare(ctx context.Context, args []string, options entities.SystemUnshareOptions) error {
+	unshare := func() error {
+		cmd := exec.Command(args[0], args[1:]...)
+		cmd.Env = unshareEnv(ic.Libpod.StorageConfig().GraphRoot, ic.Libpod.StorageConfig().RunRoot)
+		cmd.Stdin = os.Stdin
+		cmd.Stdout = os.Stdout
+		cmd.Stderr = os.Stderr
+		return cmd.Run()
+	}
+
+	if options.RootlessCNI {
+		rootlesscni, err := ic.Libpod.GetRootlessCNINetNs(true)
+		if err != nil {
+			return err
+		}
+		defer rootlesscni.Cleanup(ic.Libpod)
```

Comment:
Is there a risk that we wipe out the rootless CNI namespace if we have two unshare sessions both joined to it, and one exits before the other?

Reply:
Yes that is problematic. What would be the proper way to fix this? Cleanup currently checks that there are no running containers with bridge networking and that `var/lib/cni/results` is empty. Otherwise cleanup is skipped.

Reply:
Hm. We could try reference counting, but I'm concerned we could get into a state where we never clean up if a `podman unshare` gets SIGKILL'd before it can indicate it's done.

The easiest way is probably to guarantee there's only 1 `podman unshare` shell with `--rootless-cni` active at a time using a lockfile. I don't think we'll run into cases where more than 1 session is needed very often?

Reply:
Yeah I think we should ensure that we always do cleanup. Otherwise we will not only leak the namespace but also the slirp process.

Only allowing one session feels a little odd. The only thing that will happen to the other unshare session is that we lose network connectivity because we kill the slirp process. The namespace stays alive until the process exits only the bind mount is removed bind cleanup.  Also as long as at least one container with CNI networking is running we can safely use as many sessions as we want.

I think we can live with this problem.

Reply:
OK, that ought to be fine, then.

Reply:
could we just join the network namespace without creating it?

That would also solve the issue that we don't have to cleanup on exit?

Reply:
We could do this but it doesn't solve the problem. We could skip the cleanup here but if all running cni containers are stopped we will perform the cleanup and then all running unshare sessions would lose network connectivity.
Also the reason I choose to create the ns here is because this should open up more use cases. I thought about using this to create the ns and do some setup stuff, e.g. create a interface used by cni.

Reply:
could we split the two cases, perhaps `--rootless-cni=create|join`?  I think having an option to just join the network namespace, and lose connectivity if there are no other containers running is still a valid use case.

Reply:
There is only one rootless cni namespace per user. We will always join if it exists.